### ISSUE TITLE
Microphone not working (Lenovo Thinkbook 13S, CML-U)

### ISSUE THREAD

**Issue by phocean** [thesofproject/linux#2460]
The microphone is not working with my laptop : Lenovo Thinkbook 13S.

Distribution : openSUSE Tumbleweed (20200829)

```
~ ❯ uname -a
Linux thinkpad 5.8.4-1-default #1 SMP Wed Aug 26 10:53:09 UTC 2020 (64fe492) x86_64 x86_64 x86_64 GNU/Linux


~ ❯ inxi -A                               
Audio:
  Device-1: Intel driver: sof-audio-pci 
  Sound Server: ALSA v: k5.8.4-1-default 

~ ❯ dmesg | grep snd
[   13.184060] snd_hda_intel 0000:00:1f.3: DSP detected with PCI class/subclass/prog-if info 0x040100
[   13.184064] snd_hda_intel 0000:00:1f.3: Digital mics found on Skylake+ platform, using SOF driver
[   13.649953] snd_soc_skl 0000:00:1f.3: DSP detected with PCI class/subclass/prog-if info 0x040100
[   13.649957] snd_soc_skl 0000:00:1f.3: Digital mics found on Skylake+ platform, using SOF driver
[   14.062251] snd_hda_codec_realtek ehdaudio0D0: autoconfig for ALC257: line_outs=1 (0x14/0x0/0x0/0x0/0x0) type:speaker
[   14.062254] snd_hda_codec_realtek ehdaudio0D0:    speaker_outs=0 (0x0/0x0/0x0/0x0/0x0)
[   14.062256] snd_hda_codec_realtek ehdaudio0D0:    hp_outs=1 (0x21/0x0/0x0/0x0/0x0)
[   14.062258] snd_hda_codec_realtek ehdaudio0D0:    mono: mono_out=0x0
[   14.062259] snd_hda_codec_realtek ehdaudio0D0:    inputs:
[   14.062261] snd_hda_codec_realtek ehdaudio0D0:      Mic=0x19
[   14.108320] snd_hda_codec_realtek ehdaudio0D0: ASoC: sink widget AIF1TX overwritten
[   14.108327] snd_hda_codec_realtek ehdaudio0D0: ASoC: source widget AIF1RX overwritten
[   14.108400] skl_hda_dsp_generic skl_hda_dsp_generic: snd-soc-dummy-dai <-> DMIC 6 mapping ok
[   14.108408] skl_hda_dsp_generic skl_hda_dsp_generic: snd-soc-dummy-dai <-> DMIC16kHz 7 mapping ok
[   14.108421] skl_hda_dsp_generic skl_hda_dsp_generic: snd-soc-dummy-dai <-> HDA Analog 0 mapping ok
[   14.108433] skl_hda_dsp_generic skl_hda_dsp_generic: snd-soc-dummy-dai <-> HDA Digital 1 mapping ok
[   14.108442] skl_hda_dsp_generic skl_hda_dsp_generic: snd-soc-dummy-dai <-> HDMI1 3 mapping ok
[   14.108449] skl_hda_dsp_generic skl_hda_dsp_generic: snd-soc-dummy-dai <-> HDMI2 4 mapping ok
[   14.108459] skl_hda_dsp_generic skl_hda_dsp_generic: snd-soc-dummy-dai <-> HDMI3 5 mapping ok
```

Using `arecord` results in no sound, except a high pitch noise that last less than 1 sec at the begining.

```
/tmp ❯ arecord -l
**** Liste des Périphériques Matériels CAPTURE ****
carte 1: sofhdadsp [sof-hda-dsp], périphérique 0: HDA Analog (*) []
  Sous-périphériques: 1/1
  Sous-périphérique #0: subdevice #0
carte 1: sofhdadsp [sof-hda-dsp], périphérique 1: HDA Digital (*) []
  Sous-périphériques: 1/1
  Sous-périphérique #0: subdevice #0
carte 1: sofhdadsp [sof-hda-dsp], périphérique 6: DMIC (*) []
  Sous-périphériques: 1/1
  Sous-périphérique #0: subdevice #0
carte 1: sofhdadsp [sof-hda-dsp], périphérique 7: DMIC16kHz (*) []
  Sous-périphériques: 1/1
  Sous-périphérique #0: subdevice #0

/tmp ❯ arecord -Dhw:1,6 -c2 -r48000 -fS32_LE -d 10 mytest.wav
Capture WAVE 'mytest.wav' : Signed 32 bit Little Endian, Fréquence 48000 Hz, Stéréo

^CInterrompu par le signal Interrupt...

/tmp 7s ❯ open mytest.wav
```

Attachments:

[alsa-info.txt](https://github.com/thesofproject/linux/files/5257154/alsa-info.txt)

[dmesg.txt](https://github.com/thesofproject/linux/files/5257155/dmesg.txt)
[
Downstream bug report](https://bugzilla.kernel.org/show_bug.cgi?id=209145)

**Comment by plbossart** [comment]
@phocean could you provide a screenshot of what the waveform looks like? You can do this with audacity and screenshot. Thanks!

@singalsu @juimonen @kv2019i FYI

**Comment by phocean** [comment]
@plbossart here it is:

![image](https://user-images.githubusercontent.com/4280885/93887039-74c58200-fce6-11ea-8b1b-5288cffa2fb3.png)

Let me know if I can help further.

**Comment by plbossart** [comment]
@singalsu do we have a means to disable DC offset and ramps on the DMIC? Wondering if something goes in the weeds here.

**Comment by plbossart** [comment]
@phocean Can you make sure all switches are on in alsamixer (nothing odd muted).

Also wondering if the information on DMIC is actually correct. Can you add this in /etc/modprobe.d/hda-legacy.conf

options snd_intel_dspcfg dsp_driver=1

If the mic works with this, it means that you don't have a DMIC and your BIOS is incorrect. It's a long shot but we need to double-check this.

**Comment by phocean** [comment]
@plbossart For sure nothing has never been muted on the side of `alsamixer`, that something I checked multiple times and in all possible ways.

`options snd_intel_dspcfg dsp_driver=1` does not seem to change anything.

**Comment by plbossart** [comment]
@phocean could you try the updated firmware 1.6 (https://github.com/thesofproject/sof-bin/tree/stable-v1.6)? It's a long shot but worth double-checking. Make sure you save /lib/firmware before installing stuff.

**Comment by phocean** [comment]
@plbossart sorry for the late answer. I just tested it but no change so far...

**Comment by singalsu** [comment]
@phocean Do you still have the wav file from above recording left to attach it here? I'd like to take a closer look to what it contains (noise or perfect digital silence). Or preferably make a new recording with check done that there's nothing private accidentally recorded.

I wonder if the capture gain just very low. Those power-on start transients can correspond to jet fighter loudness like noises of real world and they are in some platforms mitigated in firmware with volume ramp and IIR filter tricks.

**Comment by phocean** [comment]
@singalsu Here it is:

[Clip 3.zip](https://github.com/thesofproject/linux/files/5390368/Clip.3.zip)

Is that ok if it's an ogg file ?

**Comment by paulstelian97** [comment]
> @singalsu Here it is:
> 
> [Clip 3.zip](https://github.com/thesofproject/linux/files/5390368/Clip.3.zip)
> 
> Is that ok if it's an ogg file ?

If it's an obviously audible issue yes. But for sound distortions we would need .wav to be able to fully understand the nature of the distortion. If it's no sound at all then ogg is okay. I recommend you get a .wav as well, just in case. (and it should be recorded as .wav, NOT just converting the ogg to wav)

**Comment by phocean** [comment]
That should be better (recorded with `arecord -vv -fdat issue.wav`) :

[issue.zip](https://github.com/thesofproject/linux/files/5391262/issue.zip)

**Comment by plbossart** [comment]
@phocean we have reports of configuration issues related to DMICs, wondering if you are in the same case as https://github.com/thesofproject/linux/issues/2551 with non-existent microphones.

Can you provide the NHLT table or run the script suggested in https://github.com/thesofproject/sof-test/pull/488?

**Comment by phocean** [comment]
@plbossart 

Is that helpful?

```
❯ sudo ./nhlt-parse.py
NHLT 1 endpoints
Link(DescriptorLength=6108, LinkType=2, InstanceID=0, VendorID=32902, DeviceID=44576, RevID=1, SubsystemID=1, DevType=1, Direction=1, VirtualBusId=0)
NHLT link type PDM
NHLT SPECIFIC_CONFIG size 3
NHLT FORMAT CONFIGS 2
NHTL: 1 dmics described
```

**Comment by singalsu** [comment]
> @phocean we have reports of configuration issues related to DMICs, wondering if you are in the same case as #2551 with non-existent microphones.
> 
> Can you provide the NHLT table or run the script suggested in [thesofproject/sof-test#488](https://github.com/thesofproject/sof-test/pull/488)?

I checked the wav file (thanks for doing the capture). It looks similar to case when no digital microphones connected or not powered.

**Comment by phocean** [comment]
It does have a mic:

https://download.lenovo.com/consumer/mobiles_pub/thinkbook_13s-14s_iwl_hmm_201905.pdf

**Comment by singalsu** [comment]
Yep, there's such in the document but no info what type of microphone it is. Has the microphone worked with your device with Windows OS, if it was shipped with it pre-installed? If it would be my device I'd open it and check the mic flex cable is properly in connector but unless familiar with electronics assembly it's a task for service. A similar known working device could be tried with a live-Ubuntu 20.04 image if you would happen to know someone who could try.

**Comment by phocean** [comment]
I went through the process of creating a Windows USB drive this morning and booting on it.

At first, it missed the sound drivers, but after installing them, it worked flawlessly.

So this is definitely not a hardware issue, but missing support in Linux.

The sound drivers can be found [there](https://pcsupport.lenovo.com/us/en/products/laptops-and-netbooks/thinkbook-series/thinkbook-13s-iwl/downloads/ds539973-realtek-audio-dolby-driver-for-windows-10-64-bit-thinkbook-13s-iwl-thinkbook-14s-iwl).

**Comment by singalsu** [comment]
Thanks @phocean , good test. @plbossart Do you know if anyone in our team has this device?

**Comment by singalsu** [comment]
@phocean Still a stretch and likely stupid question, did you try to capture from hw:0,0 "HDA Analog" ? In my CML NuC the analog microphones in the front panel appear at that device.

Edit: hw:1,0 in your device apparently.

**Comment by phocean** [comment]
@singalsu Sorry but I don't know how to test it. As for devices, I only see "sof-hda-dsp".

**Comment by singalsu** [comment]
Above shows that you have 
```
carte 1: sofhdadsp [sof-hda-dsp], périphérique 0: HDA Analog (*) []
  Sous-périphériques: 1/1
  Sous-périphérique #0: subdevice #0
```

If that's still card number 1 (usually it's 0) the capture line would be:

`arecord -vvv -Dhw:1,0 -f dat -d 10 rec.wav
`
The -vvv gives an ASCII "vu-meter"

**Comment by plbossart** [comment]
@phocean we are trying to determine where the microphones are connected
- to HDaudio codec (handled by codec driver)
- or directly to the PCH (handled by SOF firmware/driver)

**Comment by phocean** [comment]
Ok, thanks.

The result is that it gets no sound at all with the internal mic, but it would record normally with a headset plugged.

And with value 0, it says "invalid value for card".

**Comment by singalsu** [comment]
I think we need this device or find out otherwise (contact Lenovo) if the microphone clock or power connections differ from usual. If you are still OK to test more, please try to copy sof-hda-generic-4ch.tplg over sof-hda-generic-2ch.tplg and sof-hda-generic-1ch.tplg. Make backup of of those files to restore or reinstall the SOF binaries after test. Also kill of pulseaudio and temporary rename to /usr/bin/pulseaudio.disabled would be good. The rename is needed to prevent it from started again.

With 4ch DMIC configuration in place the record command  would be:

`arecord -Dhw:1,6 -f dat -c 4 -d 10 rec.wav`

The result will be a 4ch wav file. If the microphone in your device would be clocked via channels 3-4 clock this could work. But e.g. if the microphones powering is unusual this won't help. You could try also simultaneously from other shell window the previous analog microphone record test (if codec and microphone would be for some reason from same Vdd).

**Comment by singalsu** [comment]
Also, is there a microphone mute function key in the device (F4 in my ThinkPad). Try to toggle it while recording.

**Comment by singalsu** [comment]
Also into list of faint ideas to try, with the 4ch topology. Try to start camera with e.g. application "Cheese". If the microphone would share the power with camera.

**Comment by phocean** [comment]
Should I reboot after overwriting the firmware files ?

I didn't and nothing changes after following all your suggestions: still no capture except the high pitch noise at the begining.

**Comment by paulstelian97** [comment]
> Should I reboot after overwriting the firmware files ?
> 
> I didn't and nothing changes after following all your suggestions: still no capture except the high pitch noise at the begining.

The other way is to rmmod and insmod SOF so that it loads the new files instead of the old ones. But yes, the simplest way is rebooting.

**Comment by phocean** [comment]
Ok, it works that way !

Normally, I would get:

```
> arecord -vv -Dhw:1,6 -f dat -c 4 -d 10 rec.wav
Recording WAVE 'rec3.wav' : Signed 16 bit Little Endian, Rate 48000 Hz, Channels 4
arecord: set_params:1349: Channels count non available
```

It would start with only 2 channels but recording no sound:

```
arecord -vv -Dhw:1,6 -f dat -c 2 -d 10 rec.wav
```

After overwriting the files, it works flawlessly.

What does it mean ? Can I do something to keep this configuration to work ?

**Comment by paulstelian97** [comment]
I think you can mark the linux-firmware package on hold until the next release of the firmware is out and ready.

**Comment by phocean** [comment]
@paulstelian97 No, because even if the microphone works, everything else (speakers) is broken. What are these files and what kind of hack can I do to have both the mic and the speakers to work?

**Comment by singalsu** [comment]
The topology overwrite should not impact playback. There's difference only in DMIC configuration in those topologies. Did you disable pulseaudio by renaming it? Then renaming (back to /usr/bin/pulseaudio and reboot) should bring you back the speakers.

Since this helped it looks like in your device the microphones are clocked via channels 3-4 HW. Are there any DMIC pins related settings in the BIOS that could change this to normal (via GPIO pins  muxing)?

We'd really need to get a device like this. If confirmed the driver should support mode to enable only channels 1-2 but start all clocks. Currently there's no such so all channels need to be captured even if there's less connected microphones in the device.

**Comment by phocean** [comment]
Of course, I set pulseaudio back. I really loose the sound after the operation. Pulseaudio does not find any device and set a virtual device.

Regarding the PINS, there isn't such a setting in the BIOS (which is very similar to Thinkpad BIOS).

**Comment by singalsu** [comment]
@juimonen Do you know if pulseaudio version used can somehow break (speakers) from mic channels count increased from 2 to 4?

**Comment by juimonen** [comment]
@phocean can you attach here a recording from the arecord (with some audio to the mics) so we could see what channels get occupied? Also pulseaudio starting log would be good.

**Comment by gbrdead** [comment]
I am willing to take the role of a reporter. I have the same laptop (presumably).

I am using Debian Testing.

1. If the package firmware-sof-signed is not installed then there is no sound card detected.

2. After I installed firmware-sof-signed (https://packages.debian.org/bullseye/firmware-sof-signed), output worked fine but the microphone did not record any sound.

3. After I copied `/lib/firmware/intel/sof-tplg-v1.6.1/sof-hda-generic-4ch.tplg` over `sof-hda-generic-2ch.tplg` and restarted (without disabling pulseaudio), the following command:

`arecord -vv -Dhw:0,7 -f dat -c 4 rec.wav`

produced the first evidence after half an year that this computer actually has a microphone inside! Here is the file: https://www.voidland.org/.mic/rec.wav . The noise in the beginning is not from me, of course.

Here is the output of arecord -l:

```
card 0: sofhdadsp [sof-hda-dsp], device 0: HDA Analog (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: sofhdadsp [sof-hda-dsp], device 1: HDA Digital (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: sofhdadsp [sof-hda-dsp], device 6: DMIC (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: sofhdadsp [sof-hda-dsp], device 7: DMIC16kHz (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```


How can I make a GUI application (e.g. firefox) to use hw0,7? I am using KDE.

**Comment by gbrdead** [comment]
> How can I make a GUI application (e.g. firefox) to use hw0,7? I am using KDE.

I managed to do this:
1. Add the following two lines in `/etc/pulse/default.pa`:
```
load-module module-alsa-source device=hw:0,7
set-default-source alsa_input.hw_0_7
```
2. Restart and use the input named "Comet Lake PCH-LP cAVS" (without "Multichannel") in KMix.

**Comment by paulstelian97** [comment]
> 3. After I copied `/lib/firmware/intel/sof-tplg-v1.6.1/sof-hda-generic-4ch.tplg` over `sof-hda-generic-2ch.tplg` and restarted (without disabling pulseaudio), the following command:
> 
> `arecord -vv -Dhw:0,7 -f dat -c 4 rec.wav`
> 
> produced the first evidence after half an year that this computer actually has a microphone inside! Here is the file: https://www.voidland.org/.mic/rec.wav . The noise in the beginning is not from me, of course.

Yikes that's a long codec initialization time (for the noise part).

@juimonen Any clue how to compensate for the fact that this noise exists? @singalsu Does the volume component have an adjustable startup delay/volume curve?

**Comment by singalsu** [comment]
@gbrdead Seems that your device has two microphones those are connected to channels 3 and 4. There's no microphone in channels 1 and 2 and the not-connected microphones create the start transient that's seen.  (non-toggling PDM bus appears as full scale DC and then the stepped HW gain ramp and high-pass create the effect seen)  

![image](https://user-images.githubusercontent.com/28754793/107928814-78052700-6f81-11eb-86a7-b556e6890e77.png)

Device hw:0.7 gives 16 kHz sampled audio, you should have same success with hw:0,6 that is configured for 48 kHz. The start transient is longer for 16 kHz mode. There may be a way to improve it (move start transient mitigation totally to FW side since the HW mitigation settles quite slow) but please use the primary 48 kHz device if that works for you.

I need to test if accessing ch3-4 connected microphones can be fixed with a custom topology to appear as normal 2ch device. But since the BIOS information for microphones is incorrect for this device it won't owrk out-of-box. There will be need to copy the custom topology over the sof-hda-generic-2ch.tplg.

**Comment by singalsu** [comment]
Based on the observation I created this pull request https://github.com/thesofproject/sof/pull/3847 . @gbrdead Can you please try this topology:

[sof-hda-generic-2ch-pdm1.zip](https://github.com/thesofproject/linux/files/5981682/sof-hda-generic-2ch-pdm1.zip)

Unzip and copy sof-hda-generic-2ch-pdm1.tplg over /lib/firmware/intel/sof-tplg/sof-hda-generic-2ch.tplg. I just tested with my 4 mic device (Lenovo X1) that these commands captured microphones 3 and 4 as stereo.

```
arecord -Dhw:0,6 -f dat -d 10 rec06.wav
arecord -Dhw:0,7 -f dat -r 16000 -d 10 rec07.wav

```
Since alsamixer view changes from 4ch back to 2ch check that Dmic0 and Dmic1 gains and mute switches are sane.

**Comment by kv2019i** [comment]
@gbrdead To make the topology selection automatic, can you help to capture the full NHLT dump from the laptop in question:

sudo cat /sys/firmware/acpi/tables/NHLT >NHLT.bin

... and attach "NHLT.bin" to this bug. If the FW fix in https://github.com/thesofproject/linux/issues/2460#issuecomment-779212719 help, we need to figure out how the kernel could automatically load this topology on affected systems. We basicly have two options:
  1) we add an exception based on DMI name to soc-acpi-intel-cml-match.c (each affected product listed separately), or
  2) if the routing is visible in NHLT data, we could have generic logic that works on all systems where BIOS can describe the configuration

FYI for @plbossart

**Comment by gbrdead** [comment]
The microphone works much better with the patched topology:
1. Higher volume and sensitivity.
2. No noise in the beginning of the recording.

The modification to /etc/pulse/default.pa is still necessary for audacity et al.

[NHLT.zip](https://github.com/thesofproject/linux/files/5983815/NHLT.zip)

**Comment by singalsu** [comment]
@gbrdead Nice that it worked, now the low-level fix is known and need to figure out how to solve the upper level things. Can you still describe what is the issue with 48 kHz hw:0,6 device. Is the sound muted? Do you get errors from arecord?

The gains can be set via command line and alsamixer for 48 kHz device

```
$ amixer cset name='Dmic0 Capture Volume' 50,50
numid=49,iface=MIXER,name='Dmic0 Capture Volume'
  ; type=INTEGER,access=rw---R--,values=2,min=0,max=70,step=0
  : values=50,50
  | dBscale-min=-50.00dB,step=1.00dB,mute=1

$ amixer cset name='Dmic0 Capture Switch' on,on
numid=50,iface=MIXER,name='Dmic0 Capture Switch'
  ; type=BOOLEAN,access=rw------,values=2
  : values=on,on

```
For 16 kHz device there's control

```
$ amixer cset name='Dmic1 2nd Capture Volume' 50,50
numid=52,iface=MIXER,name='Dmic1 2nd Capture Volume'
  ; type=INTEGER,access=rw---R--,values=2,min=0,max=70,step=0
  : values=50,50
  | dBscale-min=-50.00dB,step=1.00dB,mute=1

```

**Comment by gbrdead** [comment]
Both hw:0,6 and hw0,7 work well with the patched topology.

**Comment by plbossart** [comment]
> 1. we add an exception based on DMI name to soc-acpi-intel-cml-match.c (each affected product listed separately), or
> 2. if the routing is visible in NHLT data, we could have generic logic that works on all systems where BIOS can describe the configuration

I re-checked the NHLT document: https://01.org/sites/default/files/595976_intel_sst_nhlt.pdf

I don't see anything that would tell us how the microphones are connected. I only see channels and mic array geometry, nothing on which PDM wires are used. 

It probably works in Windows because the mic configuration is hidden in binary blobs, so there's nothing that's exposed. It's actually a good example of missing information, the only way by which we could select the topology dynamically would be by parsing NHLT blobs.

Instead of a topology switch, could we alternatively have a Kcontrol?

**Comment by singalsu** [comment]
@plbossart This can be decoded from binary part of the blob, the blob that was shared here has the FIFO packer set up differently than in various 2ch NHLTs we quickly collected with @kv2019i , also we checked that 4ch can be detected from there. The PCM data FIFOs source is set to pdm1 instead of normal pdm0 in Thinkbook13S. It's then another matter if want to to decode this in the driver since we have different revisions of HW and the needed bit-fields to decode.

**Comment by plbossart** [comment]
@singalsu do we only have 2 configurations for 2ch? 

Also what would happen e.g. in the 1ch case, would we need to provide 4 topologies?

**Comment by singalsu** [comment]
@plbossart Yep, the same connection issue can be made with single microphone. So with the naming that I used in my SOF PR there could be need for sof-hda-generic-1ch-pdm1.tplg too.

**Comment by plbossart** [comment]
> @plbossart Yep, the same connection issue can be made with single microphone. So with the naming that I used in my SOF PR there could be need for sof-hda-generic-1ch-pdm1.tplg too.

how would we pick ch0 or ch1 on that link though?

**Comment by singalsu** [comment]
@plbossart There's no need to select ch0 or ch1 in mono case, the single microphone data appears on both slots. It's available in firmware IPC (swapped mono) but it's not used currently.

We don't actually use mono PCM format for 1ch dmic. The capture data appears with the current topologies as dual mono. There's an unsolved DMA configuration issue with mono DMIC data.

**Comment by shining** [comment]
> I managed to do this:
> 
>     1. Add the following two lines in `/etc/pulse/default.pa`:
> 
> 
> ```
> load-module module-alsa-source device=hw:0,7
> set-default-source alsa_input.hw_0_7
> ```
> 

Hello,

We have the following device on Levono P14s :
```
00:1f.3 Audio device [0403]: Intel Corporation Comet Lake PCH-LP cAVS [8086:02c8] (prog-if 80)
```

The internal mic works almost out of the box on Archlinux with kernel 5.11, after installing the sof firmware 1.7.

On Debian 10, it required more steps : 

1. installing 5.10 kernel from backport + sof firmware 1.6 from testing (or even the last 1.7 version manually, not sure)
2. options snd-intel-dspcfg dsp_driver=3 to select SOF
3. make hw_0_7 the default like suggested above (or hw_0_6?)

Could anyone explain why steps 2 and 3 are needed on Debian, and how comes it work out of the box on Archlinux ?

Also it looks like we lost the HDMI output option after step 3.

Thanks a lot for the help.

**Comment by plbossart** [comment]
> On Debian 10, it required more steps :
> 
> 1. installing 5.10 kernel from backport + sof firmware 1.6 from testing (or even the last 1.7 version manually, not sure)
> 2. options snd-intel-dspcfg dsp_driver=3 to select SOF
> 3. make hw_0_7 the default like suggested above (or hw_0_6?)
> 
> Could anyone explain why steps 2 and 3 are needed on Debian, and how comes it work out of the box on Archlinux ?

no idea @shining, the autodetection of SOF should work the same. Assuming that you are testing the two distros on the SAME hardware, we should use the same information to select SOF v. legacy HDaudio.

Please try without step2 and provide the detailed dmesg log with the following file added to /etc/modprobe.d/sof-dyndbg.conf 
[sof-dyndbg.conf.txt](https://github.com/thesofproject/linux/files/6288438/sof-dyndbg.conf.txt)

**Comment by shining** [comment]
It was two laptops from my colleagues that should have exactly the same specs.
I just received mine which comes with Windows 10 preinstalled. Is there anything useful to get from it (firmware and stuff) before replacing it by Debian ?

**Comment by paulstelian97** [comment]
> It was two laptops from my colleagues that should have exactly the same specs.
> I just received mine which comes with Windows 10 preinstalled. Is there anything useful to get from it (firmware and stuff) before replacing it by Debian ?

Should just be able to install Debian directly. Ensure you have backup media (f.e. don't bust the recovery partition) so you can return to Windows in the future if you need to.

**Comment by Elinvynia** [comment]
Update from Fedora 34 which switches to pipewire and promises to fix some sound issues in the kernel: broken by default without the patched topology file mentioned above.
Relevant fedora discussion: https://ask.fedoraproject.org/t/internal-microphone-does-not-work-on-thinkbook-13s-iml/12881/2

**Comment by shining** [comment]
> > On Debian 10, it required more steps :
> > 
> > 1. installing 5.10 kernel from backport + sof firmware 1.6 from testing (or even the last 1.7 version manually, not sure)
> > 2. options snd-intel-dspcfg dsp_driver=3 to select SOF
> > 3. make hw_0_7 the default like suggested above (or hw_0_6?)
> > 
> > Could anyone explain why steps 2 and 3 are needed on Debian, and how comes it work out of the box on Archlinux ?
> 
> no idea @shining, the autodetection of SOF should work the same. Assuming that you are testing the two distros on the SAME hardware, we should use the same information to select SOF v. legacy HDaudio.
> 
> Please try without step2 and provide the detailed dmesg log with the following file added to /etc/modprobe.d/sof-dyndbg.conf
> [sof-dyndbg.conf.txt](https://github.com/thesofproject/linux/files/6288438/sof-dyndbg.conf.txt)

Looks like step 2 was not needed on my machine, so the autodetection of SOF works indeed.

What about step 3 ?

I still don't undestand what the HDA Analog and HDA Digital devices are for, and why there are 4 capture devices ?

On my other laptop there is only one recording device and it works :
```
**** List of CAPTURE Hardware Devices ****
card 1: PCH [HDA Intel PCH], device 0: ALC3232 Analog [ALC3232 Analog]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```

**Comment by plbossart** [comment]
> > > 1. installing 5.10 kernel from backport + sof firmware 1.6 from testing (or even the last 1.7 version manually, not sure)
> > > 2. options snd-intel-dspcfg dsp_driver=3 to select SOF
> > > 3. make hw_0_7 the default like suggested above (or hw_0_6?)

> Looks like step 2 was not needed on my machine, so the autodetection of SOF works indeed.
> 
> What about step 3 ?
>

I don't know what 'make hw_0_7' refers to.  Never used it.

> I still don't undestand what the HDA Analog and HDA Digital devices are for, and why there are 4 capture devices ?

Yes, that may be a bit confusing. Some HDaudio devices have an analog and a SPDIF digital output, but we can't know this at the SOF level. So we expose two separate streams, knowing that one of them may or may not work.

> On my other laptop there is only one recording device and it works :
> 
> ```
> **** List of CAPTURE Hardware Devices ****
> card 1: PCH [HDA Intel PCH], device 0: ALC3232 Analog [ALC3232 Analog]
>   Subdevices: 1/1
>   Subdevice #0: subdevice #0
> ```

Can you share more information on what the other recording devices are? e.g. share the results of alsa-info.

Also if you have a Lenovo Thinkbook 13S, as discussed above there is a configuration issue with a PDM0/PDM1 inversion that no one else has done, and that means that to get working audio you need to override the default topology file by the one shared above, see https://github.com/thesofproject/linux/issues/2460#issuecomment-779212719

**Comment by shining** [comment]
> I don't know what 'make hw_0_7' refers to. Never used it.
> 

See https://github.com/thesofproject/linux/issues/2460#issuecomment-778805723

> 
> Yes, that may be a bit confusing. Some HDaudio devices have an analog and a SPDIF digital output, but we can't know this at the SOF level. So we expose two separate streams, knowing that one of them may or may not work.
> 

Ok so looks like you confirm that it's necessary to configure pulse default depending on the hardware for now ?
No hope for a solution ?

> 
> Can you share more information on what the other recording devices are? e.g. share the results of alsa-info.
> 

Exactly the same as in the first post, except it's card 0 and not card 1.

> Also if you have a Lenovo Thinkbook 13S

It's a P14s, overriding the default topology file does not seem necessary. 

[alsa-info.txt.S6gge5ZXkr.gz](https://github.com/thesofproject/linux/files/6327642/alsa-info.txt.S6gge5ZXkr.gz)
[sof-debug.log.gz](https://github.com/thesofproject/linux/files/6327643/sof-debug.log.gz)

**Comment by plbossart** [comment]
> > I don't know what 'make hw_0_7' refers to. Never used it.
> 
> See [#2460 (comment)](https://github.com/thesofproject/linux/issues/2460#issuecomment-778805723)

why is this necessary? if you use alsa-ucm-conf, you don't need to tweak PulseAudio. the Linux Arch information seems to be completely outdated.
> 
> > Yes, that may be a bit confusing. Some HDaudio devices have an analog and a SPDIF digital output, but we can't know this at the SOF level. So we expose two separate streams, knowing that one of them may or may not work.
> 
> Ok so looks like you confirm that it's necessary to configure pulse default depending on the hardware for now ?

not at all, this information seems completely wrong. Device 0,7 is the DMIC at 16kHz, not the one you want at all

Your local mic is device hw:0,6 and the headset mic is device hw:0,0

card 0: sofhdadsp [sof-hda-dsp], device 0: HDA Analog (*) []
card 0: sofhdadsp [sof-hda-dsp], device 6: DMIC (*) []

> No hope for a solution ?

well most folks have no issues with HDaudio + DMIC, we've worked hard to make it turn key and most likely your distro uses obsolete bits.

please see https://thesofproject.github.io/latest/getting_started/index.html#debug-audio-issues-on-intel-platforms to check for yourself what might be wrong.

**Comment by shining** [comment]
Yes, as I explained, this is on Debian 10.
To get local mic working I already had to install kernel and sof firmware from testing (future Debian 11).
alsa-ucm-conf also only exists in testing and this looks harder to backport.

So it's awesome if everything works well already with the future Debian 11, I will test it to confirm, and let's hope it will be released this year.
But I would also would like to know what's the best workaround for Debian 10 for now.

What about the following then ?
```
load-module module-alsa-source device=hw:0,6
set-default-source alsa_input.hw_0_6
```
This will maybe force the local mic all the time, even when a headset mic is plugged ?
But still a good workaround for now ?

In any cases, thanks a lot for the help, it's starting to be clearer, and thanks a lot for your work.

**Comment by shining** [comment]
I confirm it works well with Debian 11, without any hack. :+1: 

The default mic switches from local mic to headset mic when I plug a headset and vice versa.
This can be seen in pavucontrol or with the following command : 
`pacmd list-sources | grep -e 'index:' -e device.string -e 'name:'`

**Comment by sunweaver** [comment]
> Based on the observation I created this pull request [thesofproject/sof#3847](https://github.com/thesofproject/sof/pull/3847) . @gbrdead Can you please try this topology:
> 
> [sof-hda-generic-2ch-pdm1.zip](https://github.com/thesofproject/linux/files/5981682/sof-hda-generic-2ch-pdm1.zip)
> 
> Unzip and copy sof-hda-generic-2ch-pdm1.tplg over /lib/firmware/intel/sof-tplg/sof-hda-generic-2ch.tplg. I just tested with my 4 mic device (Lenovo X1) that these commands captured microphones 3 and 4 as stereo.
> 
> ```
> arecord -Dhw:0,6 -f dat -d 10 rec06.wav
> arecord -Dhw:0,7 -f dat -r 16000 -d 10 rec07.wav
> ```
> 
> Since alsamixer view changes from 4ch back to 2ch check that Dmic0 and Dmic1 gains and mute switches are sane.

@singalsu I just took your topology file and replaced sof-hda-generic-2ch.tplg on a freshly installed Debian 11 system. Test device is a Lenovo 13s (which my school customer has plenty of).

Without your topology file, the internal "Digital Microphone" is shown in Pulseaudio and alsamixer, but I cannot record sound with the internal mic. Neither do I see any input level in pavucontrol for the "Digital Microphone".

With your binary topology file offered for download further up in thesofproject/linux#2460, I can start using the microphone.

So for the Lenovo Thinkpad 13S, this bug is definitely still an issue in Debian 11.

About your PR, I saw that upstream closed the PR when renaming the "master" branch to "main" and it seems like noone ever cared about re-opening it.

So, I wonder what the status of this problem currently is in SOF upstream and how we (I am a Debian Developer) can get SOF HDA DSP working for the Lenovo Thinkpad 13S in Debian 11.

Thanks for your work on this! Looking forward to some extra info.

**Comment by plbossart** [comment]
@sunweaver the status has not changed: the SOF driver as currently implemented cannot know which PDM interface is used. Most platforms use PDM0 but Lenovo for some reason used PDM1. Unfortunately this information is set in the 'topology file' and we just don't know how to select this topology.

Rather than adding multiple topology files and try to figure out what platform needs what topology, we wanted to change the approach and instead of using the information stored in a topology file we could use the information provided in the platform firmware (DSDT/NHLT). I believe @singalsu did quite a bit of work on this already at the firmware level, this will hopefully come in a future SOF release.

**Comment by johnnrandall** [comment]
I've got a similar issue to @phocean above. My laptop is also a Thinkbook 13.

My outputs are below:

uname -a
Linux lenovo 5.4.0-100-generic #113-Ubuntu SMP Thu Feb 3 18:43:29 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

inxi -A
Audio:     Device-1: Intel driver: sof-audio-pci 
           Sound Server: ALSA v: k5.4.0-100-generic 

dmesg | grep snd
[    4.973474] snd_hda_intel 0000:00:1f.3: DSP detected with PCI class/subclass/prog-if info 0x040100
[    4.973699] snd_hda_intel 0000:00:1f.3: Digital mics found on Skylake+ platform, using SOF driver
[    5.369091] snd_hda_codec_realtek ehdaudio0D0: autoconfig for ALC257: line_outs=1 (0x14/0x0/0x0/0x0/0x0) type:speaker
[    5.369103] snd_hda_codec_realtek ehdaudio0D0:    speaker_outs=0 (0x0/0x0/0x0/0x0/0x0)
[    5.369111] snd_hda_codec_realtek ehdaudio0D0:    hp_outs=1 (0x21/0x0/0x0/0x0/0x0)
[    5.369116] snd_hda_codec_realtek ehdaudio0D0:    mono: mono_out=0x0
[    5.369120] snd_hda_codec_realtek ehdaudio0D0:    inputs:
[    5.369127] snd_hda_codec_realtek ehdaudio0D0:      Mic=0x19
[    5.417987] snd_hda_codec_realtek ehdaudio0D0: ASoC: sink widget AIF1TX overwritten
[    5.418028] snd_hda_codec_realtek ehdaudio0D0: ASoC: source widget AIF1RX overwritten
[    5.419118] skl_hda_dsp_generic skl_hda_dsp_generic: snd-soc-dummy-dai <-> DMIC 6 mapping ok
[    5.419300] skl_hda_dsp_generic skl_hda_dsp_generic: snd-soc-dummy-dai <-> DMIC16kHz 7 mapping ok
[    5.419556] skl_hda_dsp_generic skl_hda_dsp_generic: snd-soc-dummy-dai <-> HDA Analog 0 mapping ok
[    5.419784] skl_hda_dsp_generic skl_hda_dsp_generic: snd-soc-dummy-dai <-> HDA Digital 1 mapping ok
[    5.419944] skl_hda_dsp_generic skl_hda_dsp_generic: snd-soc-dummy-dai <-> HDMI1 3 mapping ok
[    5.420100] skl_hda_dsp_generic skl_hda_dsp_generic: snd-soc-dummy-dai <-> HDMI2 4 mapping ok
[    5.420247] skl_hda_dsp_generic skl_hda_dsp_generic: snd-soc-dummy-dai <-> HDMI3 5 mapping ok

The headphone microphone works but the internal one has never shown any kind of input level.

Any assistance would be gratefully received.

**Comment by plbossart** [comment]
@johnnrandall "Linux lenovo 5.4.0-100-generic" is likely rather too old for proper SOF support. It's probably one of the first versions we upstreamed. Please consider updating your system if possible, it's very difficult to solve problems on kernels that are 12 versions old. Thanks for your understanding.

**Comment by johnnrandall** [comment]
Which kernel do you recommend? I've tried a more recent one but unfortunately that alone didn't resolve it which is why I reverted to this one.

**Comment by plbossart** [comment]
5.10+ is recommended.

The solution for Lenovo devices is to try to use the topology provided in https://github.com/thesofproject/linux/files/5981682/sof-hda-generic-2ch-pdm1.zip

Unzip and copy sof-hda-generic-2ch-pdm1.tplg over /lib/firmware/intel/sof-tplg/sof-hda-generic-2ch.tplg (KEEP THE INITIAL VERSION AS A BACKUP).

**Comment by johnnrandall** [comment]
I tried the link and followed the instructions. And yes it was the solution. I can hardly contain my thankfulness! On this basis I will stick with my current kernel unless you recommend otherwise. Thank you once again, and have a great evening.

**Comment by plbossart** [comment]
@johnnrandall glad it works for you. At some point we may be able to do this switch automagically but for now we have to override this topology file. You can update the kernel to a more recent version, it will not change anything. The one thing you need to watch out for is that distribution updates may override this file, so you may have to redo this override if the mic becomes silent again.

**Comment by plbossart** [comment]
There will be no magical solution to this problem, the only work-around is to load an alternate topology. This will be enabled starting with kernel 5.20 and SOF 2.3. see details of the documentation here: https://thesofproject.github.io/latest/getting_started/intel_debug/suggestions.html#digital-mic-issues

**Comment by unoexperto** [comment]
Folks, I'm on Lenovo T14 AMD Gen4. I don't have *.tplg files on the disk at all. Could you please advise what to do ?

My stats:

```
uname -a
Linux expert-thinkpad 6.8.0-31-generic #31-Ubuntu SMP PREEMPT_DYNAMIC Sat Apr 20 00:40:06 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

inxi -A
Audio:
  Device-1: AMD Rembrandt Radeon High Definition Audio driver: snd_hda_intel
  Device-2: AMD ACP/ACP3X/ACP6x Audio Coprocessor driver: snd_pci_acp6x
  Device-3: AMD Family 17h/19h HD Audio driver: snd_hda_intel
  API: ALSA v: k6.8.0-31-generic status: kernel-api
  Server-1: PipeWire v: 1.0.5 status: active

```

**Comment by plbossart** [comment]
@unoexperto that's a different platform, please file a separate issue for AMD to support
